In [1]:
import pandas as pd

# Direct Marketing Campaigns of Portuguese Banking Clients
2 years worth of data (2008-2010).
4,119 records.
20 independent variables.  
1 dependent variable.

**Note from the authors of the dataset:**
> Missing Attribute Values: There are several missing values in some categorical attributes, all coded with the "unknown" label. These missing values can be treated as a possible class label or using deletion or imputation techniques.

Will define the NA value when reading in the dataset.

In [2]:
banking = pd.read_csv('./data/bank-additional.csv', header=0, sep=";", na_values="unknown")
banking.shape

(4119, 21)

In [3]:
banking.head().T

0            1            2            3  \
age                      30           39           25           38   
job             blue-collar     services     services     services   
marital             married       single      married      married   
education          basic.9y  high.school  high.school     basic.9y   
default                  no           no           no           no   
housing                 yes           no          yes          NaN   
loan                     no           no           no          NaN   
contact            cellular    telephone    telephone    telephone   
month                   may          may          jun          jun   
day_of_week             fri          fri          wed          fri   
duration                487          346          227           17   
campaign                  2            4            1            3   
pdays                   999          999          999          999   
previous                  0            0            0            0   
poutcome        nonexistent  nonexistent  nonexistent  nonexistent   
emp.var.rate           -1.8          1.1          1.4          1.4   
cons.price.idx       92.893       93.994       94.465       94.465   
cons.conf.idx         -46.2        -36.4        -41.8        -41.8   
euribor3m             1.313        4.855        4.962        4.959   
nr.employed          5099.1         5191       5228.1       5228.1   
y                        no           no           no           no   

                                4  
age                            47  
job                        admin.  
marital                   married  
education       university.degree  
default                        no  
housing                       yes  
loan                           no  
contact                  cellular  
month                         nov  
day_of_week                   mon  
duration                       58  
campaign                        1  
pdays                         999  
previous                        0  
poutcome              nonexistent  
emp.var.rate                 -0.1  
cons.price.idx               93.2  
cons.conf.idx                 -42  
euribor3m                   4.191  
nr.employed                5195.8  
y                              no

In [4]:
banking.describe(include="all")

age     job  marital          education default housing  loan  \
count   4119.000000    4080     4108               3952    3316    4014  4014   
unique          NaN      11        3                  7       2       2     2   
top             NaN  admin.  married  university.degree      no     yes    no   
freq            NaN    1012     2509               1264    3315    2175  3349   
mean      40.113620     NaN      NaN                NaN     NaN     NaN   NaN   
std       10.313362     NaN      NaN                NaN     NaN     NaN   NaN   
min       18.000000     NaN      NaN                NaN     NaN     NaN   NaN   
25%       32.000000     NaN      NaN                NaN     NaN     NaN   NaN   
50%       38.000000     NaN      NaN                NaN     NaN     NaN   NaN   
75%       47.000000     NaN      NaN                NaN     NaN     NaN   NaN   
max       88.000000     NaN      NaN                NaN     NaN     NaN   NaN   

         contact month day_of_week  ...     campaign        pdays  \
count       4119  4119        4119  ...  4119.000000  4119.000000   
unique         2    10           5  ...          NaN          NaN   
top     cellular   may         thu  ...          NaN          NaN   
freq        2652  1378         860  ...          NaN          NaN   
mean         NaN   NaN         NaN  ...     2.537266   960.422190   
std          NaN   NaN         NaN  ...     2.568159   191.922786   
min          NaN   NaN         NaN  ...     1.000000     0.000000   
25%          NaN   NaN         NaN  ...     1.000000   999.000000   
50%          NaN   NaN         NaN  ...     2.000000   999.000000   
75%          NaN   NaN         NaN  ...     3.000000   999.000000   
max          NaN   NaN         NaN  ...    35.000000   999.000000   

           previous     poutcome emp.var.rate  cons.price.idx  cons.conf.idx  \
count   4119.000000         4119  4119.000000     4119.000000    4119.000000   
unique          NaN            3          NaN             NaN            NaN   
top             NaN  nonexistent          NaN             NaN            NaN   
freq            NaN         3523          NaN             NaN            NaN   
mean       0.190337          NaN     0.084972       93.579704     -40.499102   
std        0.541788          NaN     1.563114        0.579349       4.594578   
min        0.000000          NaN    -3.400000       92.201000     -50.800000   
25%        0.000000          NaN    -1.800000       93.075000     -42.700000   
50%        0.000000          NaN     1.100000       93.749000     -41.800000   
75%        0.000000          NaN     1.400000       93.994000     -36.400000   
max        6.000000          NaN     1.400000       94.767000     -26.900000   

          euribor3m  nr.employed     y  
count   4119.000000  4119.000000  4119  
unique          NaN          NaN     2  
top             NaN          NaN    no  
freq            NaN          NaN  3668  
mean       3.621356  5166.481695   NaN  
std        1.733591    73.667904   NaN  
min        0.635000  4963.600000   NaN  
25%        1.334000  5099.100000   NaN  
50%        4.857000  5191.000000   NaN  
75%        4.961000  5228.100000   NaN  
max        5.045000  5228.100000   NaN  

[11 rows x 21 columns]

*Please note, according to the dataset description, a client that had not been contacted previously is represented with the value 999 in the "contact" column.*

In [5]:
banking.isna().sum()

age                 0
job                39
marital            11
education         167
default           803
housing           105
loan              105
contact             0
month               0
day_of_week         0
duration            0
campaign            0
pdays               0
previous            0
poutcome            0
emp.var.rate        0
cons.price.idx      0
cons.conf.idx       0
euribor3m           0
nr.employed         0
y                   0
dtype: int64

Will drop all records that contain any null values since we need to reduce our dataset to <= 3,000 records.

In [6]:
banking_nonulls = banking.dropna(how='any')
banking_nonulls.shape

(3090, 21)

In summary, we:
- imported our original 4,119 record dataset
- defined & identified the null values present
- removed the rows with any null value present

This has lead to a reduction of 803 records.  I recommend we either keep the dataset as is (3,090) or remove 90 more records in order to meet the recommended dataset size.  If we decide to keep dataset as is, we can export `banking_nonulls` dataframe to a CSV.  If we decide to remove more, please make any more removals as you see fit before exporting 


In [7]:
banking_nonulls.to_csv('./data/cleaned-bank.csv')